lark/parse_tree_builder.py

```diff
@@ -31,40 +32,52 @@ def __call__(self, children):
         # local reference to Tree.meta reduces number of presence checks
         if isinstance(res, Tree):
             res_meta = res.meta
-            for c in children:
-                if isinstance(c, Tree):
-                    child_meta = c.meta
-                    if not child_meta.empty:
-                        res_meta.line = child_meta.line
-                        res_meta.column = child_meta.column
-                        res_meta.start_pos = child_meta.start_pos
-                        res_meta.empty = False
-                        break
-                elif isinstance(c, Token):
-                    res_meta.line = c.line
-                    res_meta.column = c.column
-                    res_meta.start_pos = c.pos_in_stream
-                    res_meta.empty = False
-                    break
-
-            for c in reversed(children):
-                if isinstance(c, Tree):
-                    child_meta = c.meta
-                    if not child_meta.empty:
-                        res_meta.end_line = child_meta.end_line
-                        res_meta.end_column = child_meta.end_column
-                        res_meta.end_pos = child_meta.end_pos
-                        res_meta.empty = False
-                        break
-                elif isinstance(c, Token):
-                    res_meta.end_line = c.end_line
-                    res_meta.end_column = c.end_column
-                    res_meta.end_pos = c.end_pos
-                    res_meta.empty = False
-                    break
+
+            src_meta = self._pp_get_meta(children)
+            if src_meta is not None:
+                res_meta.line = src_meta.line
+                res_meta.column = src_meta.column
+                res_meta.start_pos = src_meta.start_pos
+                res_meta.empty = False
+
+            src_meta = self._pp_get_meta(reversed(children))
+            if src_meta is not None:
+                res_meta.end_line = src_meta.end_line
+                res_meta.end_column = src_meta.end_column
+                res_meta.end_pos = src_meta.end_pos
+                res_meta.empty = False
 
         return res
 
+    def _pp_get_meta(self, children):
+        for c in children:
+            if isinstance(c, Tree):
+                if not c.meta.empty:
+                    return c.meta
+            elif isinstance(c, Token):
+                return c
+
+class PropagatePositions_IgnoreWs(PropagatePositions):
+    def _pp_get_meta(self, children):
```

Comment:
We can prevent duplicating the entire function if the base function reads a single class attribute (e.g. `if self.dont_ignore_ws or (c and not c.isspace())`. The child class would then set the attribute to True. But I am not sure if it is better.

Reply:
It might be a bit shorter, but this feels more readable to me (I actually tried your version first, but didn't like it). Also the change would make the default mode slightly slower, though that hardly matters. Anyway, it's all the same to me, so if you want to change it later you're welcome to.